plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java

```diff
@@ -301,4 +317,225 @@ private void terminate(ILaunch launch, Collection<IStatus> errors) {
 	private Stream<ILaunch> getLaunches() {
 		return Stream.of(launchManager.getLaunches());
 	}
+
+	public void startDebugging(IServer server, IProgressMonitor monitor) throws CoreException {
+		IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+		if (dc == null) {
+			throw toCoreException(NLS.bind("Could not find deployment config was for {0}. "
+					+ "Your server adapter refers to an inexistant service"
+					+ ", there are no pods for it "
+					+ "or there are no labels on those pods pointing to the wanted deployment config.", 
+					server.getName()));
+		}
+
+		DebuggingContext debugContext = getDebuggingContext(dc);
+		int remotePort = debugContext.getDebugPort();
+		if( remotePort == -1 ) {
+			debugContext.setDebugPort(8787);//TODO get default port from server settings?
+		}
+		IDebugListener listener = new IDebugListener() {
+
+			public void onDebugChange(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				if (debuggingContext.getPod() == null) {
+					throw toCoreException("Unable to connect to remote pod");
+				}
+				int localPort = mapPortForwarding(debuggingContext, monitor);
+
+				if(isJavaProject(server)) {
+					ILaunch debuggerLaunch = attachRemoteDebugger(server, localPort, monitor);
+					if( debuggerLaunch != null ) {
+						overrideHotcodeReplace(server, debuggerLaunch);
+					}
+				}
+			}
+
+			@Override
+			public void onPodRestart(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				onDebugChange(debuggingContext, monitor);
+			}
+		};
+		debugContext.setDebugListener(listener);
+		enableDebugMode(dc, debugContext, monitor);
+	}
+
+	/**
+	 * Used in
+	 * 1) OpenShiftLaunchController.launch() - in a mode other then debug.
+	 * 2) OpenShiftShutdownController.stop() - if current mode was debug.
+	 *
+	 * @param server
+	 * @param dc
+	 * @param debugContext
+	 * @param monitor
+	 * @throws CoreException
+	 */
+	public void stopDebugging(IServer server, IProgressMonitor monitor)
+			throws CoreException {
+		IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+		if(dc == null) {
+			//something is wrong, do as much as possible.
+			terminateRemoteDebugger(server);
+			return;
+		}
+		DebuggingContext debugContext = getDebuggingContext(dc);
+
+		IDebugListener listener = new IDebugListener() {
+
+			public void onDebugChange(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+				terminateRemoteDebugger(server);
+				unMapPortForwarding(debuggingContext.getPod());
+			}
+
+			@Override
+			public void onPodRestart(DebuggingContext debuggingContext, IProgressMonitor monitor)
+					throws CoreException {
+			}
+		};
+		debugContext.setDebugListener(listener);
+		disableDebugMode(dc, debugContext, monitor);
```

Comment:
So, for instance, DEV_MODE = false is not equivalent to absence of this key, and even if it is false, the presence of DEBUG_PORT makes the state different from when it is absent? 
Then, at enabling debug mode, we have to remember the exact values for affected keys in IDeploymentConfig and restore them at disabling debug. Is it correct? I will try to do that.